src/rgw/rgw_bucket_sync.cc

```diff
@@ -0,0 +1,878 @@
+                                   const rgw_zone_id& _zone,
+                                   std::optional<rgw_bucket> _bucket,
+                                   const rgw_sync_policy_group& group,
+                                   rgw_sync_data_flow_group *_default_flow,
+                                   std::set<rgw_zone_id> *_pall_zones,
+                                   CB filter_cb) {
+  zone = _zone;
+  bucket = _bucket;
+  default_flow = _default_flow;
+  pall_zones = _pall_zones;
+
+  rgw_sync_bucket_entity zb(zone, bucket);
+
+  status = group.status;
+
+  std::vector<rgw_sync_bucket_pipes> zone_pipes;
+
+  string bucket_key = (bucket ? bucket->get_key() : "*");
+
+  /* only look at pipes that touch the specific zone and bucket */
+  for (auto& pipe : group.pipes) {
+    if (pipe.contains_zone_bucket(zone, bucket)) {
+      ldout(cct, 20) << __func__ << "(): pipe_map (zone=" << zone << " bucket=" << bucket_key << "): adding potential pipe: " << pipe << dendl;
+      zone_pipes.push_back(pipe);
+    }
+  }
+
+  const rgw_sync_data_flow_group *pflow;
+
+  if (!group.data_flow.empty()) {
+    pflow = &group.data_flow;
+  } else {
+    if (!default_flow) {
+      return;
+    }
+    pflow = default_flow;
+  }
+
+  auto& flow = *pflow;
+
+  pall_zones->insert(zone);
+
+  /* symmetrical */
+  if (flow.symmetrical) {
+    for (auto& symmetrical_group : *flow.symmetrical) {
+      if (symmetrical_group.zones.find(zone) != symmetrical_group.zones.end()) {
+        for (auto& z : symmetrical_group.zones) {
+          if (z != zone) {
+            pall_zones->insert(z);
+            try_add_source(z, zone, zone_pipes, filter_cb);
+            try_add_dest(zone, z, zone_pipes, filter_cb);
+          }
+        }
+      }
+    }
+  }
+
+  /* directional */
+  if (flow.directional) {
+    for (auto& rule : *flow.directional) {
+      if (rule.source_zone == zone) {
+        pall_zones->insert(rule.dest_zone);
+        try_add_dest(zone, rule.dest_zone, zone_pipes, filter_cb);
+      } else if (rule.dest_zone == zone) {
+        pall_zones->insert(rule.source_zone);
+        try_add_source(rule.source_zone, zone, zone_pipes, filter_cb);
+      }
+    }
+  }
+}
+
+/*
+ * find all relevant pipes in our zone that match {dest_bucket} <- {source_zone, source_bucket}
+ */
+vector<rgw_sync_bucket_pipe> rgw_sync_group_pipe_map::find_source_pipes(const rgw_zone_id& source_zone,
+                                                                        std::optional<rgw_bucket> source_bucket,
+                                                                        std::optional<rgw_bucket> dest_bucket) const {
+  vector<rgw_sync_bucket_pipe> result;
+
+  auto range = find_pipes(sources, source_zone, source_bucket);
+
+  for (auto iter = range.first; iter != range.second; ++iter) {
+    auto pipe = iter->second;
+    if (pipe.dest.match_bucket(dest_bucket)) {
+      result.push_back(pipe);
+    }
+  }
+  return std::move(result);
```

Comment:
i see several compiler warnings from return statements like this:
```
warning: moving a local object in a return statement prevents copy elision [-Wpessimizing-move]
```
that can be resolved with `return result;`

Reply:
https://github.com/yehudasa/ceph/commit/58403d570de3d966c0a99a3ef4f0ce1d904f7654